### ISSUE TITLE
[bitnami/thanos] Add field for serviceaccount annotations

### ISSUE THREAD

**Issue by Shavakan** [bitnami/charts#2009]
**Which chart**:

Thanos

**Description**

Thanos v0.11.0 introduced Web Identity providers for IAM credentials for AWS EKS (https://github.com/thanos-io/thanos/pull/2033). To make use of it, one should add annotations on serviceaccounts that look like this: `eks.amazonaws.com/role-arn: arn:aws:iam::123456789012:role/eksctl-irptest-addon-iamsa-default-my-serviceaccount-Role1-UCGG6NDYZ3UE` ([AWS Documentation](https://aws.amazon.com/ko/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/)) 

I'd like to request a field in thanos serviceaccounts for such annotations.

**Steps to reproduce the issue:**

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.1.0", GitCommit:"b29d20baf09943e134c2fa5e1e1cab3bf93315fa", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-19T13:57:45Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by dani8art** [comment]
Hi @Shavakan , thank you for your suggestion, in which component would you need this annotation? I mean, we would need to add these annotations to `storegateway`, `ruler`, `querier`, `compactor` and `bucketweb` or just for one of them? I guess you are talking about `storegateway` right? what would you actually like to aim adding these annotations?

**Comment by Shavakan** [comment]
My apologies for the lack of clarity. These annotations are needed when using AWS S3 as object storage on AWS EKS. According to the official [Thanos component diagram](https://thanos.io/img/arch.jpg), that would be `storegateway`, `ruler`, `compactor`, and `bucketweb` (which is not on the diagram, but, by its functionality to represent tsdb block data in human readable form, presumably needs object storage access).

**Comment by dani8art** [comment]
If I understand well, the services that need theses service account annotations are the ones that will need access to AWS services, right? In the case of Thanos, it would be Minio, wouldn't be? As I understand adding annotations to `storegateway`, `ruler`, `compactor` or `bucketweb` wouldn't have any effect since they aren't going to interact with AWS services, don't you think so? 

Regarding that, what we would need is adding the service account to Minio which is going to be the one that will access AWS services (S3), right?

Or in this case, you are not using minio but S3 directly and then `storegateway`, `ruler`, `compactor` and `bucketweb` are going to be connecting to AWS services (S3)

In that case, what would be your proposal? I can see two way of doing that:

1. Annotations per component:
```
storegateway:
  serviceAccount:
    annotations:
      eks.amazonaws.com/role-arn: arn:aws:iam::123456789012:role/eksctl-irptest-addon-iamsa-default-my-serviceaccount-Role1-UCGG6NDYZ3UE
ruler:
  serviceAccount:
    annotations:
      eks.amazonaws.com/role-arn: arn:aws:iam::123456789012:role/eksctl-irptest-addon-iamsa-default-my-serviceaccount-Role1-UCGG6NDYZ3UE
# each service that needs the annotations
```

2. Object Storage access control then the annotations declare here will be added to each service that needs it:
```
objectStorage:
  accessControl:
    annotations:
      eks.amazonaws.com/role-arn: arn:aws:iam::123456789012:role/eksctl-irptest-addon-iamsa-default-my-serviceaccount-Role1-UCGG6NDYZ3UE
```

**Comment by javsalgar** [comment]
Hi,

So that I understand well, do we need different service accounts per component or would the same work for all of them?

**Comment by javsalgar** [comment]
Ok, I just checked the chart. We create a service account per component, so first it would be great if @Shavakan could confirm which the specific one that needs the iam binding is

**Comment by Shavakan** [comment]
As per @dani8art 's second suggestion, I couldn't figure out where the `objectStorage` block would be used. Did you mean objectstoreConfig? If so then to my knowledge, the object storage configuration yaml file is ingested by [thanos](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/cmd/thanos/flags.go#L52) and does not have real impact on k8s serviceaccount annotations. Correct me if I'm mistaken :)

So I took his first approach and manually added IAM binding annotations on the aforementioned components. I do agree that it is redundant to add the annotations to four different components though. A better pattern would be to have this annotation be applied to all required serviceaccounts, but I failed to find a good place for this. This is because I don't really see the need for differentiating IAM roles for each thanos components so far.

**Comment by dani8art** [comment]
> A better pattern would be to have this annotation be applied to all required serviceaccounts 

That is what I wanted to mean when I said `objectStorage` a common point where we declare annotations for all the components that need to connect to `S3`. Maybe as you say we should find a better name for that to be consistent with Thanos naming.

**Comment by dani8art** [comment]
Hi @juan131, could you add your thoughts here, please?

**Comment by Reuuke** [comment]
Hi.
I just added support to declare custom annotations to service accounts of any Thanos components. 
The solution that also solves this issue.

**Comment by dani8art** [comment]
That's cool, thank you so much. We will continue the discussion there.

**Comment by Shavakan** [comment]
Thanks @Reuuke and @dani8art for your help!
I will now close this issue

**Comment by Dhanabalan-pop** [comment]
Does this is added in the code?

**Comment by alvneiayu** [comment]
hi @Dhanabalan-pop

The change is already included in our last version 5.1.1

https://github.com/bitnami/charts/blob/4d08038088fc3d3fdafec7cae7c2c2ded760adab/bitnami/thanos/templates/ruler/serviceaccount.yaml#L11
https://github.com/bitnami/charts/blob/4d08038088fc3d3fdafec7cae7c2c2ded760adab/bitnami/thanos/templates/storegateway/serviceaccount.yaml#L11
https://github.com/bitnami/charts/blob/4d08038088fc3d3fdafec7cae7c2c2ded760adab/bitnami/thanos/templates/bucketweb/serviceaccount.yaml#L11
https://github.com/bitnami/charts/blob/4d08038088fc3d3fdafec7cae7c2c2ded760adab/bitnami/thanos/templates/query/serviceaccount.yaml#L12
https://github.com/bitnami/charts/blob/4d08038088fc3d3fdafec7cae7c2c2ded760adab/bitnami/thanos/templates/compactor/serviceaccount.yaml#L11

Thanks a lot

Álvaro
